### PR TITLE
Add ProcessInfo to process generators

### PR DESCRIPTION
**Pull request by mpowelson**

### CODE REVIEW

**Comment by mpowelson** [comment]
@Levi-Armstrong What do you think about this. I did the discrete process as an example

**Comment by mpowelson** [comment]
I think one issue is that these names are not currently unique. I could just stick them in a vector.

**Comment by mpowelson** [comment]
Alright. I think this should work. I changed it to a vector and added a threadsafe container. Now each ProcessInput should have a shared_ptr to the same container.

**Comment by Levi-Armstrong** [comment]
@mpowelson So we can do this. The task hash does not change after setting the work. I think is we add `std::size_t task_hash` to process input we will have what is need for the ProcessInfo. This will also allow us to dump a graph that is annotated with the ProcessInfo having the hash value.  Let me know what you think.

``` c++
  tf::Task t = taskflow.placeholder();
  std::cout << t.hash_value() << std::endl;
  t.work([=]() { *(input.cnt) = 10; });
  std::cout << t.hash_value() << std::endl;
```

Update graph taskflow:
before: 
``` c++
tf::Task task = taskflow->emplace(node.process->generateTask(input)).name(node.process->getName());
```
after:
``` c++
tf::Task task = taskflow.placeholder();
ProcessInput pi = input;
pi.task_hash = task.hash_value();
tf::Task task = taskflow->emplace(node.process->generateTask(pi)).name(node.process->getName());
```

**Comment by mpowelson** [comment]
@Levi-Armstrong The obvious thing here is that the ProcessInput is now unique to the task. That was true before as well, but they were just direct copies. You could also do this by changing ProcessGenerator to be 
```
std::function<int()> generateConditionalTask(ProcessInput input, std::size_t unique_id) 
```

I'm not sure which is better.

**Comment by Levi-Armstrong** [comment]
> @Levi-Armstrong The obvious thing here is that the ProcessInput is now unique to the task. That was true before as well, but they were just direct copies. You could also do this by changing ProcessGenerator to be
> 
> ```
> std::function<int()> generateConditionalTask(ProcessInput input, std::size_t unique_id) 
> ```
> 
> I'm not sure which is better.

I like this approach. Lets do it this way.

**Comment by mpowelson** [comment]
Which way? Changing generateConditionalTask or changing ProcessInput?

**Comment by Levi-Armstrong** [comment]
> Which way? Changing generateConditionalTask or changing ProcessInput?

Sorry, Your suggestion generateConditionalTask.

**Comment by mpowelson** [comment]
@Levi-Armstrong I added the unique ID. I think this is ready for review.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/continuous_contact_check_process_generator.h [hunk 1]
```diff
@@ -70,9 +70,20 @@ class ContinuousContactCheckProcessGenerator : public ProcessGenerator
...
+  void process(ProcessInput input, std::size_t unique_id) const;
+};
+
+class ContinuousContactCheckProcessInfo : public ProcessInfo
+{
+public:
+  ContinuousContactCheckProcessInfo(std::size_t unique_id, std::string name = "Continuous Contact Check Trajectory")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/continuous_contact_check_process_generator.h:84]: Recommend moving implementation to cpp.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/discrete_contact_check_process_generator.h [hunk 2]
```diff
@@ -70,9 +71,20 @@ class DiscreteContactCheckProcessGenerator : public ProcessGenerator
...
+  void process(ProcessInput input, std::size_t unique_id) const;
+};
+
+class DiscreteContactCheckProcessInfo : public ProcessInfo
+{
+public:
+  DiscreteContactCheckProcessInfo(std::size_t unique_id, std::string name = "Discrete Contact Check Trajectory")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/discrete_contact_check_process_generator.h:85]: Recommend moving implementation to cpp.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/fix_state_bounds_process_generator.h [hunk 3]
```diff
@@ -94,9 +94,20 @@ class FixStateBoundsProcessGenerator : public ProcessGenerator
...
+  void process(ProcessInput input, std::size_t unique_id) const;
+};
+
+class FixStateBoundsProcessInfo : public ProcessInfo
+{
+public:
+  FixStateBoundsProcessInfo(std::size_t unique_id, std::string name = "Fix State Bounds")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/fix_state_bounds_process_generator.h:108]: Recommend moving implementation to cpp.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/fix_state_collision_process_generator.h [hunk 4]
```diff
@@ -113,9 +113,20 @@ class FixStateCollisionProcessGenerator : public ProcessGenerator
...
+  void process(ProcessInput input, std::size_t unique_id) const;
+};
+
+class FixStateCollisionProcessInfo : public ProcessInfo
+{
+public:
+  FixStateCollisionProcessInfo(std::size_t unique_id, std::string name = "Fix State Collision")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/fix_state_collision_process_generator.h:127]: Recommend moving implementation to cpp.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/iterative_spline_parameterization_process_generator.h [hunk 5]
```diff
@@ -90,9 +90,18 @@ class IterativeSplineParameterizationProcessGenerator : public ProcessGenerator
...
+  void process(ProcessInput input, std::size_t unique_id) const;
+};
+class IterativeSplineParameterizationProcessInfo : public ProcessInfo
+{
+public:
+  IterativeSplineParameterizationProcessInfo(std::size_t unique_id,
+                                             std::string name = "Iterative Spline Parameterization")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/iterative_spline_parameterization_process_generator.h:104]: Recommend moving implementation to cpp.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/motion_planner_process_generator.h [hunk 6]
```diff
@@ -68,9 +68,18 @@ class MotionPlannerProcessGenerator : public ProcessGenerator
...
+  void process(ProcessInput input, std::size_t unique_id) const;
+};
+
+class MotionPlannerProcessInfo : public ProcessInfo
+{
+public:
+  MotionPlannerProcessInfo(std::size_t unique_id, std::string name = "Motion Planner Process Generator")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/motion_planner_process_generator.h:82]: Recommend moving implementation to cpp.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/profile_switch_process_generator.h [hunk 7]
```diff
@@ -79,9 +79,18 @@ class ProfileSwitchProcessGenerator : public ProcessGenerator
...
+  void process(ProcessInput input, std::size_t unique_id) const;
+};
+
+class ProfileSwitchProcessInfo : public ProcessInfo
+{
+public:
+  ProfileSwitchProcessInfo(std::size_t unique_id, std::string name = "Profile Switch")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/profile_switch_process_generator.h:93]: Recommend moving implementation to cpp.

File: tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/seed_min_length_process_generator.h [hunk 8]
```diff
@@ -70,15 +70,24 @@ class SeedMinLengthProcessGenerator : public ProcessGenerator
...
                  int subdivisions) const;
 };
+
+class SeedMinLengthProcessInfo : public ProcessInfo
+{
+public:
+  SeedMinLengthProcessInfo(std::size_t unique_id, std::string name = "Seed Min Length")
+    : ProcessInfo(unique_id, std::move(name))
+  {
+  }
```
Comment (Levi-Armstrong) [inline tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/seed_min_length_process_generator.h:89]: Recommend moving implementation to cpp.

**Review by Levi-Armstrong — COMMENTED** [review]
This was a quick turnaround +1, I have a few recommendation. I have been slowly migrating implementations in header to the cpp for classes that are templated in the hope that it will improve compile time over time. Not if it will have a great impact.

**Review by Levi-Armstrong — APPROVED** [review]

**Comment by Levi-Armstrong** [comment]
@mpowelson Once CI is happy this is good to merge. Thank you for the changes.

**Review by Levi-Armstrong — APPROVED** [review]

**Comment by mpowelson** [comment]
@Levi-Armstrong Why does the code coverage build fail sometimes?

**Comment by Levi-Armstrong** [comment]
> @Levi-Armstrong Why does the code coverage build fail sometimes?

It seems to always be because of OMPL failing to find a plan. I think it has to do with the available resource on the CI machine. If this happen a restart usually solves it but I usually just merge it if it fails for this case.

**Pull request merged**